courses/utils.py

```diff
@@ -22,3 +28,87 @@ def exception_logging_generator(generator):
 
 def is_grade_valid(override_grade: float):
     return 0.0 <= override_grade <= 1.0
+
+
+def generate_program_certificate(user, program):
+    """
+    Create a program certificate if the user has a course certificate
+    for each course in the program. Also, It will create the
+    program enrollment if it does not exist for the user.
+
+    Args:
+        user (User): a Django user.
+        program (programs.models.Program): program where the user is enrolled.
+
+    Returns:
+        (ProgramCertificate or None, bool): A tuple containing a
+        ProgramCertificate (or None if one was not found or created) paired
+        with a boolean indicating whether the certificate was newly created.
+    """
+    existing_cert_queryset = ProgramCertificate.objects.filter(
+        user=user, program=program
+    )
+    if existing_cert_queryset.exists():
+        ProgramEnrollment.objects.get_or_create(
+            program=program, user=user, defaults={"active": True, "change_status": None}
+        )
+        return existing_cert_queryset.first(), False
+
+    courses_in_program_ids = set(program.courses.values_list("id", flat=True))
+    num_courses_with_cert = (
+        CourseRunCertificate.objects.filter(
+            user=user, course_run__course_id__in=courses_in_program_ids
+        )
+        .distinct()
+        .count()
+    )
+
+    if len(courses_in_program_ids) > num_courses_with_cert:
+        return None, False
+
+    program_cert = ProgramCertificate.objects.create(user=user, program=program)
+    if program_cert:
+        log.info(
+            "Program certificate for [%s] in program [%s] is created.",
+            user.username,
+            program.title,
+        )
+        _, created = ProgramEnrollment.objects.get_or_create(
+            program=program, user=user, defaults={"active": True, "change_status": None}
+        )
+
+        if created:
+            log.info(
+                "Program enrollment for [%s] in program [%s] is created.",
+                user.username,
+                program.title,
+            )
+
+    return program_cert, True
+
+
+def get_program_certificate(enrollment):
+    """
+    Resolve a certificate for this enrollment if it exists
+    """
+    user_id = enrollment.user_id
+    if isinstance(enrollment, CourseRunEnrollment):
+        # No need to include a certificate if there is no corresponding wagtail page
+        # to support the render
+        if (
+            not enrollment.run.course.program.page
```

Comment:
It breaks the program UI if Program CMS pages don’t exist yet. It throws `courses.models.Program.page.RelatedObjectDoesNotExist: Program has no page.`

We should have a safe fallback to it just like we do for program UI.

You can reproduce this by removing the program page in CMS and loading the dashboard with enrolled program courses

Reply:
fixed